feat(helm) add namespaceOverride

TheRealNoob opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
adds support for `namespaceOverride` in values file, useful if this chart is subcharted.  For example, many people include myself probably subchart this as part of their "standard monitoring deployment" and want to direct metrics-server off to namespace=kube-system.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
none

TheRealNoob committed:
[helm] add namespace override

rexagod commented:
/ok-to-test

TheRealNoob commented:
@stevehipwell I would really like to keep moving this PR forward with your support.  I have looked into each SIG-instrumentation managed project and found that all of them which support helm charts already implement `namespaceOverride`, leaving metrics-server as the outlier which this PR addresses:
* [kube-state-metrics](https://github.com/prometheus-community/helm-charts/blob/7d6d8c024842af646a10f11cede6729f664da075/charts/kube-state-metrics/values.yaml#L477)
* [prometheus-adapter](https://github.com/prometheus-community/helm-charts/blob/7d6d8c024842af646a10f11cede6729f664da075/charts/prometheus-adapter/values.yaml#L30)

In addition to the [above projects](https://github.com/kubernetes-sigs/metrics-server/pull/1711#issuecomment-3267192527) which already implement `namespaceOverride`, I have found several other kubernetes-sig projects that do:
* [descheduler](https://github.com/kubernetes-sigs/descheduler/blob/9bfdbe92e910a2f86e84329d212f2857cfbb1893/charts/descheduler/values.yaml#L47)
* [node-feature-discovery](https://github.com/kubernetes-sigs/node-feature-discovery/blob/ab679b078bbaed7dcddf463771b0e738f86a1bc7/deployment/helm/node-feature-discovery/values.yaml#L26)
* [node-feature-discovery-operator](https://github.com/kubernetes-sigs/node-feature-discovery-operator/blob/813e4ef9aadd688ca18bd1e8acde82712c57d7fa/deploy/helm/nfd-operator/values.yaml#L15)
* [headlamp](https://github.com/kubernetes-sigs/headlamp/blob/bf37735f9bf6a96a1a7dcc76cbbed2039cf504f9/charts/headlamp/values.yaml#L26)
* [aws-ebs-csi-driver](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/ef4b3eef1573423e723d5c046d2068fb2a4f7beb/charts/aws-ebs-csi-driver/values.yaml#L460)

Because metrics-server is a staple of every monitoring stack, it should support the standard patterns operators use to deploy them.  This is a purely opt-in feature with zero impact when unset -- the rendered output is identical to today's behavior.  Lacking namespaceOverride doesn't make metrics-server safer -- it just forces users to manage it outside their unified deployment or deploy it to the wrong namespace.

I'd welcome any specific technical concerns about this approach.

stevehipwell commented:
@TheRealNoob do the charts you've linked to above have a governance process or are they free contribution? Free contribution charts generally tend towards exposing the the whole KRM API via the Helm chart values. This chart is explicitly opinionated with the aim to be the "correct" way to deploy Metrics Server with Helm. This change is not aligned with the aim of this Helm chart.

aw-talos commented:
@stevehipwell Where is the aim of the Helm chart documented?

TheRealNoob committed:
Merge branch 'master' into namespace-override

TheRealNoob commented:
I understand your concern about exposing too much of the KRM API.  However in this case, the user already has the ability to control the namespace via built-in Helm capabilities `helm install -n NAMESPACE`.  This PR only offers an alternative way to control it.  It adds no security risk, doesn't change anything for existing users (unless they opt-in), and doesn't encroach across the KRM API at all.  I have also demonstrated that this feature is an accepted standard by many organizations including prometheus-community, kubernetes (sig-scheduling, sig-node, sig-ui, sig-cloud-provider), and others.

I'd also like to understand what you mean by this chart being "explicitly opinionated". If there's a documented design philosophy for this chart, I'd like to review it so I can align this PR accordingly.  If it's an informal position, I think it's worth reconsidering given how broadly this pattern is accepted across the Kubernetes ecosystem.  Ultimately, the other projects I've listed implemented namespaceOverride because their users asked for it -- the same is true here.  An opinionated chart should still align with the expectations of the ecosystem it serves.

rexagod commented:
/assign @stevehipwell 
/triage accepted

rexagod commented:
/cc @skl

TheRealNoob commented:
@stevehipwell If you could spare some time to review please?  I really don't want to see this sit several months again with no progress.

stevehipwell commented on charts/metrics-server/templates/_helpers.tpl:
```diff
@@ -23,6 +23,17 @@ If release name contains chart name it will be used as a full name.
+  {{- if .Values.namespaceOverride -}}
+    {{- .Values.namespaceOverride -}}
+  {{- else -}}
+    {{- .Release.Namespace -}}
+  {{- end -}}
+{{- end -}}
```
```suggestion
{{- define "metrics-server.namespace" -}}
{{- .Values.namespaceOverride | default .Release.Namespace -}}
{{- end -}}
```

TheRealNoob replied:
good idea.  thanks

stevehipwell reviewed (changes requested):
@TheRealNoob just a minor change.

TheRealNoob committed:
Update charts/metrics-server/templates/_helpers.tpl

compact helper logic

Co-authored-by: Steve Hipwell <steve.hipwell@gmail.com>

skl commented:
/retest

skl commented:
@TheRealNoob ci failed a couple of times, I wonder if you need another rebase?

TheRealNoob committed:
Merge branch 'kubernetes-sigs:master' into namespace-override

TheRealNoob commented:
I merged main.  I had a look at the failing test and I suspect it's just flaky.  at the very least it's not related to this change since the test case is metrics-server's exposing CPU metrics; a e2e test.  i saw this same error when i first opened the PR then later on it went away.  perhaps run the test a few times until it passes?  or I can add blank commits to trigger it?  I'll wait for your thoughts.

TheRealNoob commented:
I see that the helm installed succeeded.  [log lines](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_metrics-server/1711/pull-metrics-server-test-e2e-helm/2060821445365207040#1:build-log.txt%3A243) 243 - 260 show it installed to namespace=kube-system

skl commented:
/retest-required

skl commented on charts/metrics-server/CHANGELOG.md:
```diff
@@ -25,6 +25,7 @@
 
 - Add chart options to secure the connection between Metrics Server and the Kubernetes API Server. ([#1288](https://github.com/kubernetes-sigs/metrics-server/pull/1288)) _@mkilchhofer_
 - Add `unhealthyPodEvictionPolicy` to the Metrics Server PDB as a user enabled feature. ([#1574](https://github.com/kubernetes-sigs/metrics-server/pull/1574)) @peterabarr
+- Add `namespaceOverride` which allows deploying to a namespace other than `.Release.namespace`. ([#1710](https://github.com/kubernetes-sigs/metrics-server/pull/1710)) @TheRealNoob
```
```suggestion
- Add `namespaceOverride` which allows deploying to a namespace other than `.Release.namespace`. ([#1711](https://github.com/kubernetes-sigs/metrics-server/pull/1711)) @TheRealNoob
```

skl reviewed (commented):
Also suggest adding `namespaceOverride` to README.md.

TheRealNoob committed:
Update charts/metrics-server/CHANGELOG.md

Co-authored-by: Stephen Lang <skl@users.noreply.github.com>

TheRealNoob committed:
add README docs

Signed-off-by: TheRealNoob <mike1118@live.com>

TheRealNoob commented:
> Also suggest adding namespaceOverride to README.md.

Good call.  Done.

stevehipwell commented on charts/metrics-server/CHANGELOG.md:
```diff
@@ -25,6 +25,7 @@
 
 - Add chart options to secure the connection between Metrics Server and the Kubernetes API Server. ([#1288](https://github.com/kubernetes-sigs/metrics-server/pull/1288)) _@mkilchhofer_
 - Add `unhealthyPodEvictionPolicy` to the Metrics Server PDB as a user enabled feature. ([#1574](https://github.com/kubernetes-sigs/metrics-server/pull/1574)) @peterabarr
+- Add `namespaceOverride` which allows deploying to a namespace other than `.Release.namespace`. ([#1711](https://github.com/kubernetes-sigs/metrics-server/pull/1711)) @TheRealNoob
```
This should be under `## [UNRELEASED]`.

TheRealNoob replied:
fixed

stevehipwell reviewed (changes requested):
It might also be worth rebasing your branch to see if it fixes the failing tests?

TheRealNoob committed:
Merge branch 'master' into namespace-override

TheRealNoob committed:
Move namespaceOverride entry to UNRELEASED in CHANGELOG.md

Addresses review comment: https://github.com/kubernetes-sigs/metrics-server/pull/1711#discussion_r3631156545

Signed-off-by: TheRealNoob <mike1118@live.com>

TheRealNoob commented:
sorry i can't easily rebase at this point, because i've done previous merges and then commits on top.  it would be messy.  i just did another merge main.  hope that's fine.

TheRealNoob commented:
hey it passed this time :)

skl commented:
/lgtm

stevehipwell commented:
/label tide/merge-method-squash
/approve
